### ISSUE TITLE
Life cycle of the user argument of lws_callbacks

### ISSUE THREAD

**Issue by PoltoS** [warmcat/libwebsockets#3236]
Is it safe to assume that if not null, the `user` argument of `lws_callbacks` points to a valid memory (allocated by LWS) for this particular session?

Is it safe to assume that this are won't change and the user code can save the pointer? (means, for the same session it will remain the same along the session).

Is it safe to assume that before one of LWS_CALLBACK_CLOSED, LWS_CALLBACK_CLIENT_CLOSED, LWS_CALLBACK_CLIENT_CONNECTION_ERROR, LWS_CALLBACK_CLOSED_HTTP event was received, this memory is accessible? Maybe there is a more specific event after which this area must not be used anymore?

Is it correct that this memory is preserved across HTTP requests if keep-alive is active.

Based on this I would like contribute to the documentation. Should I or you prefer to fix it yourself. It is very ambitious in this context!

**Comment by lws-team** [comment]
Rather than that or 'fixing' documentation it's quicker to talk about the reality.  It's not a goal that any person who is in a hurry or doesn't base off the examples can be documented into shape - this kind of person typically won't read the documentation either.

wsi can be closed and their data destroyed at any time you return to the event loop.  Since the peer can decide to hang up at any time, when you return to the event loop there is always the chance the next thing handled is your wsi being closed and deleted.

Can you save a wsi or wsi->anything pointer?  Maybe.  If your wsi is killed, you will get a CLOSE callback while it is still alive.  So long as you take advantage of that to destroy your reference copies, you can still be ok.

Nothing is preserved between transactions.  You need an object whole lifecycle represents your larger task.  SS is such a long-lived object which can contain wsi operations.

**Comment by PoltoS** [comment]
I'm not in a hurry and I read the documentation pretty well. Thank you for your answer, but it does not reply to the questions at all as you was telling about `wsi`, while I was telling about the `user` ;(

Let me rephrase it.

The documentation [`struct lws_protocols`](https://libwebsockets.org/lws-api-doc-main/html/group__Protocols-and-Plugins.html#structlws__protocols) says that each session will have an allocated long-living memory of size `per_session_data_size`. A pointer to this memory will be passed as `user` parameter each time a callback is called. And from the doc it is clear that LWS is allocating and freeing it and keeps the same for a particular session.

It looks like the application can use this memory to keep some data between callbacks. Is this a correct understanding?

At which moment this structure is freed (after which callback reason)? Is it guaranteed that such a callback will arrive before this memory is freed?

**Comment by lws-team** [comment]
For accepted server connections, which it seems you are talking about, the user data exists only in the lifecycle of the wsi.  That is why the answer is in terms of the wsi.  Keeping a copied reference to the wsi itself, or things that die with it are all under the same constraint that any wsi and its baggage such as user allocation may die at any return to the event loop.

You can use the allocated user data for what you like, but its scope is from when the incoming  connection was accepted through to the close processing for the socket.

If the user data was created, you should get a CLOSE type callback on the wsi, after which it will be destroyed.

**Comment by PoltoS** [comment]
So, using it from ESTABLISHED/ HTTP to CLOSE is safer right? I think it is important to mention in the doc. I'll meanwhile check my code, but I don't use `user` outside of WRITEABLE, RECEIVED, ESTABLISHED and CLOSE, and delete all references on CLOSE.

Another topic, I think it is worth to make three different names for global_user, protocol_user and session_user. Having the same `user` name os super confusing. I've seen many of my colleagues struggling all about this name use for three different objects. Up to you, but it would make the doc much easier to read.

**Comment by lws-team** [comment]
It's again a scope thing about member names being the same across types... another common one is "name".  It does not cause problems in reality because when you dereference it, it's clear from the object->name what you are looking at.

**Comment by PoltoS** [comment]
There is a common property "name" for all objects around use. This is common for people using humain languages. But I've never heard about a common property "user".

What's your name? Andy!
What's your user? .... (I certainly don't ask for a username here, but for a "bag" the lib allocated for you by the lib).

Again, all developers who touched LWS in our company complained on how confusing it is to have the same name for three different things. And a simple rename could fix it. But let's keep LWS for really smart people ;)

**Comment by pppaulpeter** [comment]
> I stopped using user field associated with the session I was always getting crashes. Instead I'm memorizing the wsi context and keeping the related data in our code. This is more predictable and no need to know how is lws managing this user area. Of course, on Close and Error events the memory is freed and never used after.
Hi PoltoS,

thanks for the quick reply!
ok, then it is quite strange, here is ok. 
porbably Andy know the answer. @lws-team

**Comment by lws-team** [comment]
Being a library, any use of lws is a coproduction between itself and "user code".

It's useful to think about "a crash in strlen()", it's possible your libc strlen() is buggy or bad but considering you can't make use of it without a valid NUL terminated string your code has to provide, and considering what it is and what it does, it's more likely strlen() is doing what it's supposed to and the problem is elsewhere in your code, eg, it was given a NULL pointer.  Yes there was definitely a "crash in strlen()" but it may not be anything to do with strlen. 

That is why there are so many examples, if lws is broken somewhere it should be possible to show it is broken with no or almost no extra user code, then, and using the example I should be able to reproduce the breakage.  For many kinds of problem, removing the user code completely and making the standalone example part of lws was the only way to work on it cleanly.

In turn that's why it's much better to base your code on the examples, they are supposed to work and if you meet problems, it's often simpler to try the scenario on an un- or minimally- changed example and see if it's really still broken without your code.

If you do meet problems in lws + your own code, if you suspect it's to do with lifecycle, valgrind is usually the key to understand what is really going on (and in what order).

**Comment by PoltoS** [comment]
Andy, this brings us to the original discussion of docs and specs ;) strlen has a very clear description in C89 and C11 as well as in POSIX standards. The spec explains how to use it in a predictable way. And no examples needed then.

For LWS there are many examples, but they are all showing different parts of usage not covering all options. Real spec is really missing. Such a good project with such a bad spec! ;( And I understand that LWS is hard to use by its nature (targeted on performance and minimal code size), so you had to create many examples to help people. Those finally substituted a good doc.

I know that I can use valgrind and compile LWS myself and chase the bug. But 1) my project is very big, LWS is a small part of it and valgrind is a bit problematic here. 2) even if I find the issue, at the end, I have to rely in the OS existing LWS lib (we distribute on Debian-like OS) and it will take forever to get an update. Bringing all libs with us is how we currently distribute our s/w. This is our issue, but it pushes us not to go in hard debugging of LWS and rather finding another path of implementation. What we finally did. Sorry for that.

**Comment by lws-team** [comment]
It doesn't buy me anything if you use my stuff or not.  It's in successful use in many companies and projects (some very) large and small, even that doesn't buy me anything.  It's not like it's going to miss some driveby that couldn't be arsed to run valgrind and find if his problem was in his code or not.
